// @nrwl/storybook/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storybook7Version = exports.tsNodeVersion = exports.htmlWebpackPluginVersion = exports.litVersion = exports.litHtmlVersion = exports.storybookTestRunnerVersion = exports.storybookNextAddonVersion = exports.storybookSwcAddonVersion = exports.reactNativeStorybookLoader = exports.storybookReactNativeVersion = exports.viteBuilderVersion = exports.webpack5Version = exports.urlLoaderVersion = exports.svgrVersion = exports.babelPresetTypescriptVersion = exports.babelLoaderVersion = exports.babelCoreVersion = exports.storybookVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
exports.storybookVersion = '^6.5.15';
exports.babelCoreVersion = '7.12.13';
exports.babelLoaderVersion = '8.1.0';
exports.babelPresetTypescriptVersion = '7.12.13';
exports.svgrVersion = '^6.1.2';
exports.urlLoaderVersion = '^4.1.1';
exports.webpack5Version = '^5.64.0';
exports.viteBuilderVersion = '^0.2.6';
exports.storybookReactNativeVersion = '^6.0.1-beta.11';
exports.reactNativeStorybookLoader = '^2.0.5';
exports.storybookSwcAddonVersion = '^1.1.7';
exports.storybookNextAddonVersion = '^1.6.6';
exports.storybookTestRunnerVersion = '^0.7.2';
exports.litHtmlVersion = '^2.3.1';
exports.litVersion = '^2.6.1';
exports.htmlWebpackPluginVersion = '^5.5.0';
exports.tsNodeVersion = '10.9.1';
exports.storybook7Version = '^7.0.0-rc.0';
