// @nrwl/workspace/src/generators/new/generate-preset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generatePreset = exports.addPresetDependencies = void 0;
const devkit_1 = require("@nrwl/devkit");
const presets_1 = require("../utils/presets");
const versions_1 = require("../../utils/versions");
const get_npm_package_version_1 = require("../utils/get-npm-package-version");
const path_1 = require("path");
const yargsParser = require("yargs-parser");
const child_process_1 = require("child_process");
function addPresetDependencies(host, options) {
    if (options.preset === presets_1.Preset.Apps ||
        options.preset === presets_1.Preset.Core ||
        options.preset === presets_1.Preset.Empty ||
        options.preset === presets_1.Preset.NPM) {
        return;
    }
    const { dependencies, dev } = getPresetDependencies(options);
    return (0, devkit_1.addDependenciesToPackageJson)(host, dependencies, dev, (0, path_1.join)(options.directory, 'package.json'));
}
exports.addPresetDependencies = addPresetDependencies;
function generatePreset(host, opts) {
    const parsedArgs = yargsParser(process.argv, {
        boolean: ['interactive'],
        default: {
            interactive: true,
        },
    });
    const spawnOptions = {
        stdio: 'inherit',
        shell: true,
        cwd: (0, path_1.join)(host.root, opts.directory),
    };
    const pmc = (0, devkit_1.getPackageManagerCommand)();
    const executable = `${pmc.exec} nx`;
    const args = getPresetArgs(opts);
    return new Promise((resolve, reject) => {
        (0, child_process_1.spawn)(executable, args, spawnOptions).on('close', (code) => {
            if (code === 0) {
                resolve();
            }
            else {
                const message = 'Workspace creation failed, see above.';
                reject(new Error(message));
            }
        });
    });
    function getPresetArgs(options) {
        // supported presets
        return getDefaultArgs(options);
    }
    function getDefaultArgs(opts) {
        return [
            `g`,
            `@nrwl/workspace:preset`,
            `--name=${opts.appName}`,
            opts.style ? `--style=${opts.style}` : null,
            opts.linter ? `--linter=${opts.linter}` : null,
            opts.npmScope ? `--npmScope=${opts.npmScope}` : `--npmScope=${opts.name}`,
            opts.preset ? `--preset=${opts.preset}` : null,
            opts.bundler ? `--bundler=${opts.bundler}` : null,
            opts.framework ? `--framework=${opts.framework}` : null,
            opts.docker ? `--docker=${opts.docker}` : null,
            opts.packageManager ? `--packageManager=${opts.packageManager}` : null,
            opts.standaloneApi !== undefined
                ? `--standaloneApi=${opts.standaloneApi}`
                : null,
            parsedArgs.interactive ? '--interactive=true' : '--interactive=false',
            opts.routing !== undefined ? `--routing=${opts.routing}` : null,
        ].filter((e) => !!e);
    }
}
exports.generatePreset = generatePreset;
function getPresetDependencies({ preset, presetVersion, bundler, }) {
    switch (preset) {
        case presets_1.Preset.TS:
            return { dependencies: {}, dev: { '@nrwl/js': versions_1.nxVersion } };
        case presets_1.Preset.AngularMonorepo:
        case presets_1.Preset.AngularStandalone:
            return {
                dependencies: { '@nrwl/angular': versions_1.nxVersion },
                dev: {
                    '@angular-devkit/core': versions_1.angularCliVersion,
                    typescript: versions_1.typescriptVersion,
                },
            };
        case presets_1.Preset.Express:
            return { dependencies: {}, dev: { '@nrwl/express': versions_1.nxVersion } };
        case presets_1.Preset.Nest:
            return {
                dependencies: {},
                dev: { '@nrwl/nest': versions_1.nxVersion, typescript: versions_1.typescriptVersion },
            };
        case presets_1.Preset.NextJs:
            return { dependencies: { '@nrwl/next': versions_1.nxVersion }, dev: {} };
        case presets_1.Preset.ReactMonorepo:
        case presets_1.Preset.ReactStandalone:
            return {
                dependencies: {},
                dev: {
                    '@nrwl/react': versions_1.nxVersion,
                    '@nrwl/cypress': versions_1.nxVersion,
                    '@nrwl/jest': bundler !== 'vite' ? versions_1.nxVersion : undefined,
                    '@nrwl/vite': bundler === 'vite' ? versions_1.nxVersion : undefined,
                    '@nrwl/webpack': bundler === 'webpack' ? versions_1.nxVersion : undefined,
                },
            };
        case presets_1.Preset.ReactNative:
            return { dependencies: {}, dev: { '@nrwl/react-native': versions_1.nxVersion } };
        case presets_1.Preset.Expo:
            return { dependencies: {}, dev: { '@nrwl/expo': versions_1.nxVersion } };
        case presets_1.Preset.WebComponents:
            return {
                dependencies: {},
                dev: { '@nrwl/web': versions_1.nxVersion, typescript: versions_1.typescriptVersion },
            };
        case presets_1.Preset.NodeStandalone:
            return { dependencies: {}, dev: { '@nrwl/node': versions_1.nxVersion } };
        default: {
            return {
                dev: {},
                dependencies: {
                    [preset]: presetVersion !== null && presetVersion !== void 0 ? presetVersion : (0, get_npm_package_version_1.getNpmPackageVersion)(preset),
                },
            };
        }
    }
}
