// @nrwl/linter/src/generators/workspace-rules-project/workspace-rules-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.lintWorkspaceRulesProjectSchematic = exports.lintWorkspaceRulesProjectGenerator = exports.WORKSPACE_PLUGIN_DIR = exports.WORKSPACE_RULES_PROJECT_NAME = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const js_1 = require("@nrwl/js");
const path_1 = require("path");
const workspace_lint_rules_1 = require("../../utils/workspace-lint-rules");
const versions_1 = require("nx/src/utils/versions");
const versions_2 = require("../../utils/versions");
exports.WORKSPACE_RULES_PROJECT_NAME = 'eslint-rules';
exports.WORKSPACE_PLUGIN_DIR = 'tools/eslint-rules';
function lintWorkspaceRulesProjectGenerator(tree, options = {}) {
    var _a, _b;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const { addPropertyToJestConfig, jestProjectGenerator } = (0, devkit_1.ensurePackage)('@nrwl/jest', versions_2.nxVersion);
        // Noop if the workspace rules project already exists
        try {
            (0, devkit_1.readProjectConfiguration)(tree, exports.WORKSPACE_RULES_PROJECT_NAME);
            return;
        }
        catch (_c) { }
        // Create the project, the test target is added below by the jest generator
        (0, devkit_1.addProjectConfiguration)(tree, exports.WORKSPACE_RULES_PROJECT_NAME, {
            root: exports.WORKSPACE_PLUGIN_DIR,
            sourceRoot: exports.WORKSPACE_PLUGIN_DIR,
            targets: {},
        });
        // Generate the required files
        (0, devkit_1.generateFiles)(tree, (0, path_1.join)(__dirname, 'files'), workspace_lint_rules_1.workspaceLintPluginDir, {
            tmpl: '',
            offsetFromRoot: (0, devkit_1.offsetFromRoot)(exports.WORKSPACE_PLUGIN_DIR),
            rootTsConfigPath: (0, js_1.getRelativePathToRootTsConfig)(tree, exports.WORKSPACE_PLUGIN_DIR),
        });
        /**
         * Ensure that when workspace rules are updated they cause all projects to be affected for now.
         * TODO: Explore writing a ProjectGraph plugin to make this more surgical.
         */
        const nxJson = (0, devkit_1.readNxJson)(tree);
        if ((_b = (_a = nxJson.targetDefaults) === null || _a === void 0 ? void 0 : _a.lint) === null || _b === void 0 ? void 0 : _b.inputs) {
            nxJson.targetDefaults.lint.inputs.push(`{workspaceRoot}/${exports.WORKSPACE_PLUGIN_DIR}/**/*`);
            (0, devkit_1.updateNxJson)(tree, nxJson);
        }
        // Add jest to the project and return installation task
        const installTask = yield jestProjectGenerator(tree, {
            project: exports.WORKSPACE_RULES_PROJECT_NAME,
            supportTsx: false,
            skipSerializers: true,
            setupFile: 'none',
            compiler: 'tsc',
            skipFormat: true,
        });
        (0, devkit_1.updateJson)(tree, (0, path_1.join)(workspace_lint_rules_1.workspaceLintPluginDir, 'tsconfig.spec.json'), (json) => {
            if (json.include) {
                json.include = json.include.map((v) => {
                    if (v.startsWith('src/**')) {
                        return v.replace('src/', '');
                    }
                    return v;
                });
            }
            if (json.exclude) {
                json.exclude = json.exclude.map((v) => {
                    if (v.startsWith('src/**')) {
                        return v.replace('src/', '');
                    }
                    return v;
                });
            }
            return json;
        });
        // Add swc dependencies
        (0, devkit_1.addDependenciesToPackageJson)(tree, {}, { '@swc-node/register': versions_1.swcNodeVersion, '@swc/core': versions_1.swcCoreVersion });
        // Add extra config to the jest.config.ts file to allow ESLint 8 exports mapping to work with jest
        addPropertyToJestConfig(tree, (0, devkit_1.joinPathFragments)(exports.WORKSPACE_PLUGIN_DIR, 'jest.config.ts'), 'moduleNameMapper', {
            '@eslint/eslintrc': '@eslint/eslintrc/dist/eslintrc-universal.cjs',
        });
        if (!options.skipFormat) {
            yield (0, devkit_1.formatFiles)(tree);
        }
        return installTask;
    });
}
exports.lintWorkspaceRulesProjectGenerator = lintWorkspaceRulesProjectGenerator;
exports.lintWorkspaceRulesProjectSchematic = (0, devkit_1.convertNxGenerator)(lintWorkspaceRulesProjectGenerator);
