// @nrwl/storybook/src/generators/configuration/lib/user-prompts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getStorybook6Framework = exports.getStorybook7Framework = exports.getBundler = exports.getGeneratorConfigurationOptions = void 0;
const tslib_1 = require("tslib");
const utilities_1 = require("@nrwl/storybook/src/utils/utilities");
const enquirer_1 = require("enquirer");
function getGeneratorConfigurationOptions(rawSchema) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        if (rawSchema.storybook7Configuration) {
            if (!rawSchema.storybook7UiFramework) {
                rawSchema.storybook7UiFramework = yield getStorybook7Framework();
            }
        }
        else {
            if (!rawSchema.uiFramework) {
                rawSchema.uiFramework = yield getStorybook6Framework();
            }
        }
        if ((rawSchema === null || rawSchema === void 0 ? void 0 : rawSchema.uiFramework) === '@storybook/react' ||
            (rawSchema === null || rawSchema === void 0 ? void 0 : rawSchema.uiFramework) === '@storybook/web-components') {
            if (!(rawSchema === null || rawSchema === void 0 ? void 0 : rawSchema.bundler)) {
                rawSchema.bundler = yield getBundler();
            }
        }
        return rawSchema;
    });
}
exports.getGeneratorConfigurationOptions = getGeneratorConfigurationOptions;
function getBundler() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const a = yield (0, enquirer_1.prompt)([
            {
                name: 'bundler',
                message: `Choose the builder you want to use for Storybook`,
                type: 'autocomplete',
                choices: [
                    {
                        name: 'vite',
                        message: 'Use Vite.js with the @storybook/builder-vite package',
                    },
                    {
                        name: 'webpack',
                        message: 'Use Webpack 5 with the @storybook/builder-webpack5 package',
                    },
                ],
            },
        ]);
        return a.bundler;
    });
}
exports.getBundler = getBundler;
function getStorybook7Framework() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const a = yield (0, enquirer_1.prompt)([
            {
                name: 'UiFramework7',
                message: `Choose the Storybook 7 framework that you need to use`,
                type: 'autocomplete',
                choices: [
                    ...utilities_1.Constants.uiFrameworks7.map((uiFramework7) => ({
                        name: uiFramework7,
                        message: uiFramework7,
                    })),
                ],
            },
        ]);
        return a.UiFramework7;
    });
}
exports.getStorybook7Framework = getStorybook7Framework;
function getStorybook6Framework() {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const a = yield (0, enquirer_1.prompt)([
            {
                name: 'UiFramework',
                message: `Choose the Storybook framework that you need to use`,
                type: 'autocomplete',
                choices: [
                    ...Object.entries(utilities_1.Constants.uiFrameworks).map(([_key, uiFramework]) => ({
                        name: uiFramework,
                        message: uiFramework,
                    })),
                ],
            },
        ]);
        return a.UiFramework;
    });
}
exports.getStorybook6Framework = getStorybook6Framework;
